lib/Basic/Unix/TaskQueue.inc

```diff
@@ -320,164 +324,335 @@ void TaskQueue::addTask(const char *ExecPath, ArrayRef<const char *> Args,
+/// When a task has finished executing, TaskHospice waits for it to die, takes
+/// action appropriate to the cause of death, then TaskMonitor reclaims its
+/// storage.
+class TaskMonitor {
+  std::queue<std::unique_ptr<Task>> &TasksToBeExecuted;
+  TaskMap TasksBeingExecuted;
+
+  std::vector<struct pollfd> FdsBeingPolled;
+
+  const unsigned MaxNumberOfParallelTasks;
 
-      pid_t Pid = T->getPid();
+public:
+  struct CallBacks {
+    const TaskQueue::TaskBeganCallback TaskBegan;
+    const TaskQueue::TaskFinishedCallback TaskFinished;
+    const TaskQueue::TaskSignalledCallback TaskSignalled;
+    const llvm::function_ref<void()> PolledAnFd;
+  };
 
-      if (Began) {
-        Began(Pid, T->getContext());
-      }
+private:
+  CallBacks callBacks;
 
-      PollFds.push_back({ T->getPipe(), POLLIN | POLLPRI | POLLHUP, 0 });
-      // We should also poll T->getErrorPipe(), but this introduces timing
-      // issues with shutting down the task after reading getPipe().
-      ExecutingTasks[Pid] = std::move(T);
-    }
+public:
+  TaskMonitor(std::queue<std::unique_ptr<Task>> &TasksToBeExecuted,
+              const unsigned NumberOfParallelTasks, const CallBacks &callBacks)
+      : TasksToBeExecuted(TasksToBeExecuted),
+        MaxNumberOfParallelTasks(
+            NumberOfParallelTasks == 0 ? 1 : NumberOfParallelTasks),
+        callBacks(callBacks) {}
+
+  /// Run the tasks to be executed.
+  /// \return true on error.
+  bool executeTasks();
+
+private:
+  bool isFinishedExecutingTasks() const {
+    return TasksBeingExecuted.empty() && TasksToBeExecuted.empty();
+  }
 
-    assert(!PollFds.empty() &&
-           "We should only call poll() if we have fds to watch!");
-    int ReadyFdCount = poll(PollFds.data(), PollFds.size(), -1);
-    if (ReadyFdCount == -1) {
-      // Recover from error, if possible.
-      if (errno == EAGAIN || errno == EINTR)
-        continue;
+  /// Start up tasks if we aren't already at the parallel limit, and no earlier
+  /// subtasks have failed.
+  /// \return true on error.
+  bool startUpSomeTasks();
+
+  /// \return true on error.
+  bool beginExecutingATask(Task &T);
+
+  /// Enter the task and its outputs in this TaskMonitor's data structures so
+  /// it can be polled.
+  void startPollingFdsOfTask(const Task &T);
+
+  void stopPolling(ArrayRef<int> FinishedFds);
+
+  enum class PollResult { HardError, SoftError, NoError };
+  PollResult pollTheFds();
+
+  /// \return None on error.
+  Optional<std::vector<int>> readFromReadyFdsReturningFinishedOnes();
+
+  /// Ensure that events bits returned from polling are what's expected.
+  void verifyEvents(short events) const;
+
+  void readDataIfAvailable(short events, int fd, Task &T) const;
+
+  bool didTaskHangup(short events) const;
+};
+
+/// Do everything required to clean up a task when its output has hung up.
+/// Encapsulates the knowledge of how to wait, and then how to clean up.
+class TaskHospice {
```

Comment:
This name is too clever, sorry. I'm still mildly with Graydon that there's no point in making it a class if it's just a namespace (despite the access control), and then that there's no point in making it a namespace when it's just these few things.

Reply:
(Elaboration on "too clever": this is an uncommon enough word that I have to think about what it means and how it applies and then "oh yeah, that does make sense", and I'm a native English speaker. I think that's an unnecessary barrier to entry for someone who may not know the word.)

Reply:
OK, I liked the class and the clever name, but that's probably just my own brain wiring, given that you both pushed back. I dissolved the hospice. See if this version works better, tx!